atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java

```diff
@@ -0,0 +1,319 @@
+    //  and also need to cap it at the latest bucket identifier.
+
+    // We assume here that each bucket is for a unique shard and strategy
+
+    // Do we actually need to do this? At the time we were going to load 50 buckets which were fine partitions, now
+    // they're 10 minutes long. 50 x 10 minutes is 500 minutes, so almost 10 hours worth of data each time. In a normal
+    // case where startbuckets is near the current bucket identifier, we would shortcircuit, but the algorithm for
+    // iteratively
+    // loading optimal is a little bit more complex. Is it sufficient to just load the row of the startBucket (and maybe
+    // the row after it)
+    // and return everything there? There may be no buckets (if startBucket hasn't updated yet), but this should be
+    // short lived (assuming the background task is running).
+    // This however does add a dependence on the background task actually running, however, whereas a model that keeps
+    // searching until we load n buckets or reach the end
+    // would work even if the background task does not run.
+
+    // TODO:
+    //  Do an iteration of all startMajorBucketIdentifierByShardAndStrategy
+    //  Add to set, and then decrement from maxRemainingBucketsToLoad.
+    //  If less than 0, then remove from startMajorBucketIdentifierByShardAndStrategy.
+    //  Cap the major identifier to the major equivalent of the maxBucketIdentifier.
+    //  Repeat until complete
+    public Set<SweepableBucket> getSweepableBuckets(Set<Bucket> startBuckets) {
+        List<SweepAssignedBucketsRow> rows = startBuckets.stream()
+                .flatMap(bucket -> Stream.of(
+                        SweepAssignedBucketStoreKeyPersister.INSTANCE.sweepBucketsRow(bucket),
+                        SweepAssignedBucketStoreKeyPersister.INSTANCE.nextSweepBucketsRow(bucket)))
+                .collect(Collectors.toList());
+
+        return readRows(rows);
+    }
+
+    @Override
+    public void putTimestampRangeForBucket(
+            Bucket bucket, Optional<TimestampRange> oldTimestampRange, TimestampRange newTimestampRange) {
+        Cell cell = SweepAssignedBucketStoreKeyPersister.INSTANCE.sweepBucketsCell(bucket);
+        casCell(
+                cell,
+                oldTimestampRange.map(timestampRangePersister::trySerialize),
+                timestampRangePersister.trySerialize(newTimestampRange));
+    }
+
+    @Override
+    public void deleteBucketEntry(Bucket bucket) {
+        deleteCell(SweepAssignedBucketStoreKeyPersister.INSTANCE.sweepBucketsCell(bucket));
+    }
+
+    @Override
+    public TimestampRange getTimestampRangeRecord(long bucketIdentifier) {
+        Cell cell = SweepAssignedBucketStoreKeyPersister.INSTANCE.sweepBucketRecordsCell(bucketIdentifier);
+        return readCell(cell, timestampRangePersister::tryDeserialize)
+                .orElseThrow(() -> new SafeIllegalStateException(
```

Comment:
Unlike the various put methods, where callers may handle the CAS exception (which I should probably document in the interfaces...), this one throws the exception internally rather than pass through an optional.

The reasoning was that no one should call this for a record that doesn't exist, but open to just returning an optional and forcing the caller to unwrap.